### ISSUE TITLE
Redesign access to variables and their values in Process subclasses

### ISSUE THREAD

**Issue by benbovy** [benbovy/xarray-simlab#19]
Having three different properties (i.e., `state`, `change`, `rate` and `value` as an alias of `state`) is probably a bad idea:

- Allowing only one value per variable a-priori makes more sense, especially for a very generic framework.
- Most of the time only the property `value` (or `state`) is used. In a few cases other properties are used as well, e.g., setting property `change` in `.run_step()` and updating `state` in `.finalize_step()`, but this is less common.
- I haven't encountered yet any case where more than one of these properties is needed externally (i.e., in other processes). So if we need to deal with different values of the same field (delta, rate, state) inside a process we might simply use regular attributes.
- Using these properties results in quite verbose code in `Process` subclasses.
- Inputs and snapshots are only for the property `value` (or `state`).

I therefore suggest that a variable only accepts one value. There is different options for implementing this:

1. We keep the current design, remove all properties mentioned above and instead provide one property, with a short name like `.v`.

2. Maybe it would be possible to get rid of using a specific property to access to the variable value in methods of a `Process` subclass and instead use directly the attribute to access the value, e.g.,

```python
class Foo(xsimlab.Process):
    coef = xsimlab.Variable(())
    field = xsimlab.Variable((), provided=True)

    def run_step(self, dt):
        self.field = self.coef * dt
```

Option 2 looks cleaner, actually very much like [attrs](https://github.com/python-attrs/attrs). It would require a bit of redesign work, though.

Given that we create a model by providing in a dict one or several `Process` subclasses

```python
model = xsimlab.Model({'foo': Foo})
```

it should be possible to inspect models directly by using the class attributes, e.g.,

```python
model.foo.coef
```

The trick to replace the `Variable` object by their value in Process methods (`initialize`, `run_step`, etc.) would be to create instances of each `Process` subclasses of the model every time we start a simulation. As processes in a model are ordered in a computationally consistent manner, we can just initialize the subclasses in that order so that we can properly link foreign variables.

Overall, option 2 may be actually much cleaner (and possibly more efficient) than the current design.

**Comment by benbovy** [comment]
After thinking about option 2, I'm wondering if it shouldn't be better to build this directly on top of `attrs` instead of re-implementing much of its features. It has both advantages and drawbacks, though. The main risk is adding another required dependency. One of the main advantages is that if it continues to grow in popularity, users will be more familiar with this API.

A `xsimlab` process would then look very similar to a `attrs` class:

```python
@xsimlab.process(time_dependent=False)
class Foo(object):
    coef = xsimlab.variable(()) 
    field = xsimlab.variable((), provided=True)
    bar = xsimlab.foreign(Bar, 'bar')
```

What would change:

- The class doesn't inherit anymore from a `Process` base class, instead we use a decorator
- Some important parameters defining the behavior of a Process (e.g. `time_dependent`) are given using keyword arguments of the decorator instead of a attributes in a class `Meta`. Other arguments could be available as well, like `auto_doc=True` as suggested in #3.
- Variable objects in a process are declared using functions instead of class constructors (e.g., `Variable` -> `variable`)
- All derived classes like `FloatVariable` would be removed in favour of parametrized validators (see `attrs` doc). xarray-simlab would provide a bunch of useful validators in addition to those already provided in `attrs`.
- `optional` argument in `Variable` would be replaced by the specific optional validator in available in `attrs`.

**Comment by benbovy** [comment]
Option 2 (i.e., `Process` instances attributes directly pointing to the values instead of using properties) may work in some cases, e.g., when using numpy arrays and updating these in place, but not in all cases.

So properties might still be needed. But instead of defining a property in `Variable` (i.e., `Variable.value`), properties may be defined programmatically when building process classes (i.e., with `@xsimlab.process` decorator), which would give something equivalent to, e.g.,

```Python
class Foo(object):
    @property
    def coef(self):
        return self._xsimlab_coef

    @coef.setter
    def coef(self, value)
        self._xsimlab_coef = value
```

Another option (without properties) would be to have some logic which auto-update all external references to a variable each time the latter is updated.

**Comment by benbovy** [comment]
Using properties, it might be possible to run in the setters all validator or converter functions attached to their respective variables, e.g.,

```python

class Foo(object):
    @property
    def coef(self):
        return self._xsimlab_coef

    @coef.setter
    def coef(self, value)
        # --- run validators here ---
        self._xsimlab_coef = value
```

Note on validators: as `value` can be either a scalar or an array, it might be worth to provide some default validators that support both (+ broadcasting?).

**Comment by benbovy** [comment]
Using properties would also allow some defensive programming. Property setters are indeed not needed for all variables. Read-only properties would prevent setting the value of variables where this would not be allowed (i.e., if `provided=False` and if variable is not a model input).

The main problem is that whether a variable is an input or not is determined at creation of a `Model` instance, not at creation of a process class.

**Comment by benbovy** [comment]
A possible solution to get the value of a foreign variable from an instance of the process class where the original variable is defined:

```python
class Foo(object):
    def __init___(self):
        self._xsimlab_ext_bar = None

    @property
    def bar(self):
        return self._xsimlab_ext_bar.bar
```

where `self._xsimlab_ext_bar` will correspond to the instance of `Bar` used for one simulation. We'll need to implement some logic to set similar attributes for all process classes in a model, just after these latter are instantiated at the beginning of each new simulation.

**Comment by benbovy** [comment]
A alternative name for foreign variables would be `link`:

```python
@xsimlab.process(time_dependent=False)
class Foo(object):
    coef = xsimlab.variable(()) 
    field = xsimlab.variable((), provided=True)
    bar = xsimlab.link(Bar, 'bar')
```

**Comment by benbovy** [comment]
Closed in #33
